src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala

```diff
@@ -0,0 +1,124 @@
+import play.api.libs.json.{ JsValue, Json }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+@SuppressWarnings(Array("ClassNames"))
+class MigrationTo160(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer) extends StrictLogging {
+
+  def migrate(): Future[Done] = {
+    MigrationTo160.migrateReservations(instanceRepository, persistenceStore)
+  }
+}
+
+object MigrationTo160 extends StrictLogging {
+  /**
+    * This function traverse all the instances in the ZK, and moves reservation from task to the instance level
+    *
+    * @param instanceRepository
+    * @param persistenceStore
+    * @param ctx
+    * @param mat
+    * @return
+    */
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a zookeeper store to make sure that migration will run in production environment
+      * and will be skipped in a unit test.
+      */
+    val maybeStore: Option[ZkPersistenceStore] = {
+
+      def findZkStore(ps: PersistenceStore[_, _, _]): Option[ZkPersistenceStore] = {
+        ps match {
+          case zk: ZkPersistenceStore =>
+            Some(zk)
+          case lcps: LazyCachingPersistenceStore[_, _, _] =>
+            findZkStore(lcps.store)
+          case lvcps: LazyVersionCachingPersistentStore[_, _, _] =>
+            findZkStore(lvcps.store)
+          case ltcps: LoadTimeCachingPersistenceStore[_, _, _] =>
+            findZkStore(ltcps.store)
+          case other =>
+            None
+        }
+      }
+
+      findZkStore(persistenceStore)
+    }
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+      new IdResolver[Instance.Id, JsValue, String, ZkId] {
+        override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+          ZkId(category, id.idString, version)
+
+        override val category: String = "instance"
+
+        override def fromStorageId(key: ZkId): Id = Instance.Id(key.id)
+
+        override val hasVersions: Boolean = false
+
+        override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+      }
+
+    implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+      Unmarshaller.strict {
+        case ZkSerialized(byteString) =>
+          Json.parse(byteString.utf8String)
+      }
+
+    import Reservation.reservationFormat
+    import Instance.instanceJsonReads
+
+    maybeStore.map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get(instanceId)
+        }
+        .mapConcat {
+          case Some(jsValue) =>
+            val instance = jsValue.as[Instance]
+            //prior to 1.6.0, only PVs are supported only with apps,
```

Comment:
Nit:
```
            // Prior to Marathon 1.6.0, persistent volumes are supported only with apps,
            // therefore reservation objects can only appear in app instances, and since
            // an app has only one task by definition, there is only one KV pair in a taskMap
            // object.
            //
            // We use .headOption here to handle the case of apps with no persistent volumes.
```